### DISCUSSION TITLE
calling webhook - connection refused

### DISCUSSION THREAD

**Discussion by raminderis** [kubernetes-sigs/kubebuilder#3802] [Support] [answered]
so i am following the famous kubebuilder book. all went well until i started playing with webhooks to introduce mutliple versions. 

when i try to create the cronjob it calls either mutating webhook (if v1) or conversion webhook(if v2).

but both cases i get (this one is when i try to create v1 job): 
`Error from server (InternalError): error when creating "config/samples/batch_v1_cronjob.yaml": Internal error occurred: failed calling webhook "mcronjob.kb.io": failed to call webhook: Post "https://kubebuilder-webhook-service.kubebuilder-system.svc:443/mutate-batch-tutorial-kubebuilder-io-v1-cronjob?timeout=10s": dial tcp 10.96.169.192:443: connect: connection refused`

I see the service, the end points and the kubebuilder-controller-manager pod running normally. 

Few Q:
1. am i expected to see a separate webhook pod? on this page https://book.kubebuilder.io/cronjob-tutorial/running-webhook, it says webhook pod will come up, but i only see controller-manager pod with manager and rbac containers. is this correct?
2. any help what i need to do to move past this connection refused error. on the same page at the end I see it talks about bootstrapping problem, i have a feeling somehow i need to use objectSelectors, but when i try to use them in config/webhook/manifests.yaml where WebhookConfiguration is defined, it gets removed when i run "make manifests". 

i am stuck here for more then 3 days now. Any help is greatly appreciated.

**Comment by camilamacedo86** [comment, marked as answer]
Hi @raminderis,

## Webhook Server

The webhook server typically runs as a part of the controller-manager pod. The manager container inside this pod is responsible for running both the controllers and the webhook server.  See: https://github.com/kubernetes-sigs/kubebuilder/blob/2b25adec1ffb8e811a85e8d86357adcf7213bf5b/testdata/project-v4/cmd/main.go#L159-L162

- See in the main.go that the server is created within the Manager: https://github.com/kubernetes-sigs/kubebuilder/blob/2b25adec1ffb8e811a85e8d86357adcf7213bf5b/testdata/project-v4/cmd/main.go#L104

- See the controller-runtime implementations: 

https://github.com/kubernetes-sigs/controller-runtime/blob/479b723944e34ae42c9911fe01228ff34eb5ca81/pkg/manager/manager.go#L244-L247

https://github.com/kubernetes-sigs/controller-runtime/blob/479b723944e34ae42c9911fe01228ff34eb5ca81/pkg/manager/internal.go#L270-L279

## Connection Refused Error

The "connection refused" error usually indicates that something is preventing the Kubernetes API server from communicating with your webhook service. Here are a few things you can check and try:

- Did you install the cert-manager?
- Note that typically you will need to have the cert-manager installed and the certs to use within

You can check the patch: https://github.com/kubernetes-sigs/kubebuilder/blob/master/testdata/project-v4/config/default/manager_webhook_patch.yaml

You need uncomment: https://github.com/kubernetes-sigs/kubebuilder/blob/2b25adec1ffb8e811a85e8d86357adcf7213bf5b/testdata/project-v4/config/default/kustomization.yaml#L39-L142

PS.: The e2e tests are done with webhooks, you might want to give a look at: https://github.com/kubernetes-sigs/kubebuilder/blob/master/test/e2e/v4/plugin_cluster_test.go

**Other points to check**

- Check if your cluster has not NetworkPolicies or firewalls that are blocking it
- Check the manager logs 
- Check the service and endpoints

I hope that answered your question, so I am closing this one.
However, if you need please feel free to re-open it. 



Some refs:
- https://book.kubebuilder.io/cronjob-tutorial/webhook-implementation.html
-

**Reply by smv5309** [reply]
Previously I was using an older version of kubebuilder as part of operator-sdk v1.6 and the latest cert-manager 1.7.0  and I had conversion webhooks working. Recently I have migrated the operator over to he latest kubebuilder 4.6.0 and regenerated my operator and I have run into the above mentioned issue. I have made no changes to my cert manager manifests however the newest Kubebuilder removed ca-injection and added significant stuff to config/default/kustomization.yaml.
 
 Could this cause the missmatch where my operator no longer has an endpoint for 
 
 https://x-operator-webhook-service.x-operator-system.svc:443/convert?
timeout=30s": no endpoints available for service "x-operator-webhook-service" when i try to deploy any application using my CRD's.

If not is this a different issue and if so is there a solution. Is my cert-manager perhaps not getting certificates even though i see them in kubernetes 

NAMESPACE             NAME                         READY   SECRET                AGE
x-operator-system   x-operator-metrics-certs   True    metrics-server-cert   6m29s
x-operator-system   x-operator-serving-cert    True    webhook-server-cert   6m28s
